language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/extensions/ballerina/document/BallerinaLocateSyntaxTreeUtil.java

```diff
@@ -0,0 +1,91 @@
+import io.ballerina.tools.text.LinePosition;
+import io.ballerina.tools.text.TextDocument;
+import io.ballerina.tools.text.TextRange;
+import org.ballerinalang.langserver.common.utils.CommonUtil;
+import org.eclipse.lsp4j.Range;
+
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * This is the BallerinaLocateSyntaxTreeUtil class for related utils in retrieving the location of syntax nodes
+ * for a given selection.
+ */
+public class BallerinaLocateSyntaxTreeUtil {
+    private static final String POSITION = "position";
+    private static final String LOCATE_PATH_PROPERTY = "isNodePath";
+
+    public static JsonElement mapNodePath(Range range, SyntaxTree syntaxTree, JsonElement syntaxTreeJson) {
+        syntaxTreeJson.getAsJsonObject().addProperty(LOCATE_PATH_PROPERTY, true);
+        Node node = CommonUtil.findNode(range, syntaxTree);
+
+        if (node.kind() == SyntaxKind.LIST) {
+            node = node.parent();
+        }
+        if (node == syntaxTree.rootNode()) {
+            return syntaxTreeJson;
+        }
+        return findNodePath(syntaxTreeJson, syntaxTree, node);
+    }
+
+    private static JsonElement findNodePath (JsonElement syntaxTreeJson, SyntaxTree syntaxTree, Node node) {
+        Optional<JsonElement> temp = Optional.of(syntaxTreeJson);
+
+        while (temp.isPresent()) {
+            temp = findChildren(syntaxTree, node, temp.get());
+        }
+        return syntaxTreeJson;
+    }
+
+    private static Optional<JsonElement> findChildren(SyntaxTree syntaxTree, Node node, JsonElement jsonNode) {
+        for (Map.Entry<String, JsonElement> childEntry: jsonNode.getAsJsonObject().entrySet()) {
+            if (childEntry.getValue().isJsonArray()) {
+                for (JsonElement childProps: childEntry.getValue().getAsJsonArray()) {
+                    Integer isPath = evaluateRange(syntaxTree, node, childProps,
+                            childProps.getAsJsonObject().get(POSITION).getAsJsonObject());
+                    if (isPath > 0) {
+                        return Optional.of(childProps);
+                    }
+                    if (isPath < 0) {
+                        return Optional.empty();
+                    }
+                }
+            } else if (childEntry.getValue().isJsonObject() && childEntry.getValue().getAsJsonObject().has(POSITION)) {
+                Integer isPath = evaluateRange(syntaxTree, node, childEntry.getValue(),
+                            childEntry.getValue().getAsJsonObject().get(POSITION).getAsJsonObject());
+                if (isPath > 0) {
+                    return Optional.ofNullable(childEntry.getValue());
+                }
+                if (isPath < 0) {
+                    return Optional.empty();
+                }
+            }
+        }
+        return Optional.empty();
+    }
+
+    private static Integer evaluateRange(SyntaxTree syntaxTree, Node node, JsonElement jsonNode, JsonObject position) {
```

Comment:
Shall we return boolean here?

Reply:
We are checking for 3 different conditions, in order to break the loop at the right time.